File: modules/background-service-manager/android/src/main/java/expo/modules/backgroundservicemanager/BackgroundServiceManagerModule.kt
package expo.modules.backgroundservicemanager

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.PowerManager
import android.provider.Settings
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * Expo Module for managing background services and background tasks.
 * 
 * This module provides:
 * 1. Foreground Service management (for visible background work)
 * 2. WorkManager integration (for battery-efficient periodic tasks)
 * 3. Battery optimization checks and settings
 * 
 * Google Play Store Compliance:
 * - All background work uses proper Android APIs
 * - Respects battery optimization and Doze mode
 * - Provides user-visible notifications for foreground service
 * - Uses WorkManager for deferrable background tasks
 */
class BackgroundServiceManagerModule : Module() {
    
    private val context: Context
        get() = appContext.reactContext ?: throw IllegalStateException("React context is null")
    
    override fun definition() = ModuleDefinition {
        Name("BackgroundServiceManager")
        
        // ============================================================
        // FOREGROUND SERVICE FUNCTIONS
        // ============================================================
        
        /**
         * Start the foreground service with a visible notification
         * Required for any background work that should continue when app is not visible
         */
        Function("startForegroundService") { title: String, message: String ->
            try {
                NotificationForegroundService.startService(context, title, message)
                true
            } catch (e: Exception) {
                e.printStackTrace()
                false
            }
        }
        
        /**
         * Stop the foreground service
         */
        Function("stopForegroundService") {
            try {
                NotificationForegroundService.stopService(context)
                true
            } catch (e: Exception) {
                e.printStackTrace()
                false
            }
        }
        
        /**
         * Check if foreground service is currently running
         */
        Function("isForegroundServiceRunning") {
            NotificationForegroundService.isServiceRunning()
        }
        
        // ============================================================
        // WORKMANAGER FUNCTIONS (Battery-Efficient Background Tasks)
        // ============================================================
        
        /**
         * Schedule periodic background work using WorkManager
         * Minimum interval is 15 minutes (Android system requirement)
         * 
         * @param intervalMinutes How often to run the task (minimum 15)
         * @param taskType Type of task to perform
         */
        Function("scheduleBackgroundWork") { intervalMinutes: Int, taskType: String ->
            try {
                val interval = if (intervalMinutes < 15) 15 else intervalMinutes.toLong()
                BackgroundWorker.schedulePeriodicWork(context, interval, taskType)
                true
            } catch (e: Exception) {
                e.printStackTrace()
                false
            }
        }
        
        /**
         * Cancel scheduled background work
         */
        Function("cancelBackgroundWork") {
            try {
                BackgroundWorker.cancelWork(context)
                true
            } catch (e: Exception) {
                e.printStackTrace()
                false
            }
        }
        
        /**
         * Check if background work is currently scheduled
         */
        Function("isBackgroundWorkScheduled") {
            try {
                BackgroundWorker.isWorkScheduled(context)
            } catch (e: Exception) {
                e.printStackTrace()
                false
            }
        }
        
        // ============================================================
        // BATTERY OPTIMIZATION FUNCTIONS
        // ============================================================
        
        /**
         * Check if the app is ignoring battery optimizations
         * When true, app can run more freely in background (use responsibly!)
         */
        Function("isIgnoringBatteryOptimizations") {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                val powerManager = context.getSystemService(Context.POWER_SERVICE) as PowerManager
                powerManager.isIgnoringBatteryOptimizations(context.packageName)
            } else {
                true // Not applicable on older Android versions
            }
        }
        
        /**
         * Request user to disable battery optimizations for this app
         * WARNING: Only use if absolutely necessary for your app's core functionality
         * Google Play may reject apps that abuse this
         */
        AsyncFunction("requestIgnoreBatteryOptimizations") {
            try {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                    val intent = Intent(Settings.ACTION_REQUEST_IGNORE_BATTERY_OPTIMIZATIONS).apply {
                        data = Uri.parse("package:${context.packageName}")
                        addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                    }
                    context.startActivity(intent)
                    true
                } else {
                    true // Not needed on older versions
                }
            } catch (e: Exception) {
                e.printStackTrace()
                false
            }
        }
        
        /**
         * Open battery optimization settings for the app
         * Less aggressive than requesting to ignore optimizations
         */
        AsyncFunction("openBatteryOptimizationSettings") {
            try {
                val intent = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                    Intent(Settings.ACTION_IGNORE_BATTERY_OPTIMIZATION_SETTINGS).apply {
                        addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                    }
                } else {
                    Intent(Settings.ACTION_SETTINGS).apply {
                        addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                    }
                }
                context.startActivity(intent)
                true
            } catch (e: Exception) {
                e.printStackTrace()
                false
            }
        }
        
        // ============================================================
        // UTILITY FUNCTIONS
        // ============================================================
        
        /**
         * Get Android version info (useful for conditional behavior)
         */
        Function("getAndroidVersion") {
            mapOf(
                "sdkInt" to Build.VERSION.SDK_INT,
                "release" to Build.VERSION.RELEASE,
                "codename" to Build.VERSION.CODENAME
            )
        }
        
        /**
         * Check if device is in Doze mode
         * In Doze mode, background work is heavily restricted
         */
        Function("isDeviceIdleMode") {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                val powerManager = context.getSystemService(Context.POWER_SERVICE) as PowerManager
                powerManager.isDeviceIdleMode
            } else {
                false
            }
        }
    }
}

File: modules/background-service-manager/src/BackgroundServiceManagerModule.ts
import { requireNativeModule } from 'expo-modules-core';
import type { BackgroundServiceManagerModuleType, TaskType, AndroidVersionInfo } from './BackgroundServiceManager.types';

const Native: BackgroundServiceManagerModuleType = requireNativeModule('BackgroundServiceManager');

// ============================================================
// FOREGROUND SERVICE FUNCTIONS
// ============================================================

/**
 * Start the foreground service with a visible notification.
 * This is required for any background work that should continue when the app is not visible.
 * 
 * @param title - Title for the notification
 * @param message - Message for the notification
 * @returns true if service started successfully
 */
export function startForegroundService(title: string, message: string): boolean {
  return Native.startForegroundService(title, message);
}

/**
 * Stop the foreground service
 * @returns true if service stopped successfully
 */
export function stopForegroundService(): boolean {
  return Native.stopForegroundService();
}

/**
 * Check if foreground service is currently running
 * @returns true if service is running
 */
export function isForegroundServiceRunning(): boolean {
  return Native.isForegroundServiceRunning();
}

// ============================================================
// WORKMANAGER FUNCTIONS (Battery-Efficient Background Tasks)
// ============================================================

/**
 * Schedule periodic background work using WorkManager.
 * WorkManager respects Doze mode and battery optimization settings.
 * 
 * Note: Minimum interval is 15 minutes (Android system requirement)
 * 
 * @param intervalMinutes - How often to run the task (minimum 15 minutes)
 * @param taskType - Type of task to perform (default: 'notification_check')
 * @returns true if work was scheduled successfully
 */
export function scheduleBackgroundWork(
  intervalMinutes: number,
  taskType: TaskType = 'notification_check'
): boolean {
  return Native.scheduleBackgroundWork(intervalMinutes, taskType);
}

/**
 * Cancel scheduled background work
 * @returns true if work was cancelled successfully
 */
export function cancelBackgroundWork(): boolean {
  return Native.cancelBackgroundWork();
}

/**
 * Check if background work is currently scheduled
 * @returns true if work is scheduled
 */
export function isBackgroundWorkScheduled(): boolean {
  return Native.isBackgroundWorkScheduled();
}

// ============================================================
// BATTERY OPTIMIZATION FUNCTIONS
// ============================================================

/**
 * Check if the app is ignoring battery optimizations.
 * When true, the app can run more freely in the background.
 * 
 * @returns true if app is ignoring battery optimizations
 */
export function isIgnoringBatteryOptimizations(): boolean {
  return Native.isIgnoringBatteryOptimizations();
}

/**
 * Request user to disable battery optimizations for this app.
 * 
 * ⚠️ WARNING: Only use if absolutely necessary for your app's core functionality.
 * Google Play may reject apps that abuse this permission.
 * 
 * @returns Promise that resolves to true if request was successful
 */
export function requestIgnoreBatteryOptimizations(): Promise<boolean> {
  return Native.requestIgnoreBatteryOptimizations();
}

/**
 * Open battery optimization settings for the app.
 * This is less aggressive than requesting to ignore optimizations.
 * 
 * @returns Promise that resolves to true if settings opened successfully
 */
export function openBatteryOptimizationSettings(): Promise<boolean> {
  return Native.openBatteryOptimizationSettings();
}

// ============================================================
// UTILITY FUNCTIONS
// ============================================================

/**
 * Get Android version information
 * Useful for implementing conditional behavior based on Android version
 * 
 * @returns Object containing sdkInt, release, and codename
 */
export function getAndroidVersion(): AndroidVersionInfo {
  return Native.getAndroidVersion();
}

/**
 * Check if device is in Doze mode.
 * In Doze mode, background work is heavily restricted.
 * 
 * @returns true if device is in Doze mode
 */
export function isDeviceIdleMode(): boolean {
  return Native.isDeviceIdleMode();
}

// Default export with all functions
export default {
  startForegroundService,
  stopForegroundService,
  isForegroundServiceRunning,
  scheduleBackgroundWork,
  cancelBackgroundWork,
  isBackgroundWorkScheduled,
  isIgnoringBatteryOptimizations,
  requestIgnoreBatteryOptimizations,
  openBatteryOptimizationSettings,
  getAndroidVersion,
  isDeviceIdleMode,
};

